src/main/kotlin/org/elm/ide/intentions/RecordConstructorToLiteralIntention.kt

```diff
@@ -0,0 +1,90 @@
+package org.elm.ide.intentions
+
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiWhiteSpace
+import com.intellij.psi.impl.source.tree.LeafPsiElement
+import org.elm.lang.core.psi.ElmPsiFactory
+import org.elm.lang.core.psi.ElmTypes.UPPER_CASE_IDENTIFIER
+import org.elm.lang.core.psi.elements.*
+
+/**
+ * An intention action that transforms a usage of record constructor to the corresponding record literal, i.e. converts
+ * this:
+ * ```
+ * type alias MyTypeAlias = { foo: String, bar: Int }
+ *
+ * baz = MyTypeAlias "a" 1
+ * ```
+ * to this:
+ * ```
+ * type alias MyTypeAlias = { foo: String, bar: Int }
+ *
+ * baz = { foo = "a", bar = 1 }
+ * ```
+ */
+class RecordConstructorToLiteralIntention :
+    ElmAtCaretIntentionActionBase<RecordConstructorToLiteralIntention.Context>() {
+
+    /**
+     * @param functionCall The current record constructor function call, to be replaced with a record literal.
+     * @param arguments The arguments to the record literal. Each entry in the list is a pair of strings: the first is
+     * the argument name, the second is the argument's text.
+     */
+    data class Context(val functionCall: ElmFunctionCallExpr, val arguments: List<Pair<String, String>>)
+
+    override fun getText() = "Use record literal"
+    override fun getFamilyName() = text
+
+    override fun findApplicableContext(project: Project, editor: Editor, element: PsiElement): Context? {
+        // This intention is valid in a situation like this:
+        //      foo = MyRecord "a" 1
+        // In order to be valid the caret must be in the text "MyRecord". If the caret is inside this word (or just
+        // before the M) then the PSI structure is as follows:
+        //      ElmFunctionCallExpr  (this is the record constructor function)
+        //          ElmValueExpr
+        //              ElmUpperCaseQID
+        //                  LeafPsiElement (the current element, elementType = UPPER_CASE_IDENTIFIER)
+        // But if the caret is just after the last "d" of "MyRecord" the structure is as follows:
+        //      ElmFunctionCallExpr  (this is the record constructor function)
+        //          ElmValueExpr
+        //          PsiWhiteSpace    (the current element, sibling of the ElmValueExpr above)
+        val functionCall = when {
```

Comment:
IntelliJ does have an abstraction for matching "patterns" against the Psi tree to detect certain conditions (see `com.intellij.patterns.PlatformPatterns`), but I found that API confusing. I suspect that it suffers from its Java heritage, and that had it been written for Kotlin, it would work differently. That being said, I've never used it, so if you find it useful, I'd be interested in learning more.

In this case, I would probably just search upwards using the `ancestors` sequence helper functions that you found in our project and search for an `ElmFunctionCallExpr`. Then I would return early if `funcCallExpr.target != element` (or something like that--you'll have to fiddle with it).

Reply:
I had a look at `PlatformPatterns` and I agree with you: it's not the most intuitive. I looked at just using `ancestors` the way you suggested but the problem was that the intention was then available if the caret is anywhere in the "scope" of that function (i.e. when it's inside any of the function's parameters). For example: `val test = MyTypeAlias "a" {-caret-}1`. Personally I find this a bit confusing: I think it's a bit clearer if the intention is only ever available if you're on the actual constructor function name itself (`MyTypeAlias` in this example). So here's what I came up with:
```suggestion
val functionCall =
    // First condition above...
    (element.ancestors.firstOrNull { it is ElmValueExpr && it.parent is ElmFunctionCallExpr }?.parent
    // Second condition above...
        ?: element.parent.takeIf { it is ElmFunctionCallExpr && element.prevSibling is ElmValueExpr }
        ?: return null)
            as ElmFunctionCallExpr
```
I'm actually not sure that's any clearer than the original code (the indentation IntelliJ applies also doesn't seem great), though the result seems to be the same. So my gut feeling on this is to keep the code as I originally had it. What d'you think?

Reply:
I agree that the alternative code was not much clearer.

As for the intention only being available when the cursor is on the record constructor itself, yes, totally! What I meant was that you can do it in 2 steps:

1. Search up through the ancestors for a function call
2. Use the `target` property on the `ElmFuncCallExpr` that you found in step 1 to verify that it (the "target") is the element that the intention was invoked on.